python/ray/serve/batching.py

```diff
@@ -183,58 +191,64 @@ async def _consume_func_generator(
 
         while True:
-            batch: List[_SingleRequest] = await self.wait_for_batch()
-            assert len(batch) > 0
-            self_arg = batch[0].self_arg
-            args, kwargs = _batch_args_kwargs([item.flattened_args for item in batch])
-            futures = [item.future for item in batch]
-
-            # Method call.
-            if self_arg is not None:
-                func_future_or_generator = func(self_arg, *args, **kwargs)
-            # Normal function call.
-            else:
-                func_future_or_generator = func(*args, **kwargs)
-
-            if isasyncgenfunction(func):
-                func_generator = func_future_or_generator
-                await self._consume_func_generator(func_generator, futures, len(batch))
-            else:
-                try:
-                    func_future = func_future_or_generator
-                    results = await func_future
-                    self._validate_results(results, len(batch))
-                    for result, future in zip(results, futures):
-                        # If the client has disconnected, the future is canceled.
-                        # We should only set result when the connection is still live.
-                        if not future.cancelled():
-                            future.set_result(result)
-                except Exception as e:
-                    for future in futures:
-                        future.set_exception(e)
+            try:
+                self.curr_iteration_start_time = time.time()
+                batch: List[_SingleRequest] = await self.wait_for_batch()
+                assert len(batch) > 0
+                self_arg = batch[0].self_arg
+                args, kwargs = _batch_args_kwargs(
+                    [item.flattened_args for item in batch]
+                )
+                futures = [item.future for item in batch]
+
+                # Method call.
+                if self_arg is not None:
+                    func_future_or_generator = func(self_arg, *args, **kwargs)
+                # Normal function call.
+                else:
+                    func_future_or_generator = func(*args, **kwargs)
+
+                if isasyncgenfunction(func):
+                    func_generator = func_future_or_generator
+                    await self._consume_func_generator(
+                        func_generator, futures, len(batch)
+                    )
+                else:
+                    try:
+                        func_future = func_future_or_generator
+                        results = await func_future
+                        self._validate_results(results, len(batch))
+                        for result, future in zip(results, futures):
+                            _set_result_safe(future, result)
+                    except Exception as e:
+                        for future in futures:
+                            _set_exception_safe(future, e)
```

Comment:
I think this should get moved to the top-level handler, else if we have an exception raised earlier in this function body the requests could hang

Reply:
Good callout, we've moved the top-level exception handler to a separate function per [this comment](https://github.com/ray-project/ray/pull/42631#discussion_r1465247601), and that function doesn't have access to the futures, so we can't process them there.

Instead, I refactored this `_process_batch` function, so most of the code is encapsulated within a `try-except`. That way, if an exception is raised after we get the futures, the futures will still get set.